## 예외

Java에서 문법에 맞지 않게 작성한 코드는 사전에 컴파일러에 의해 오류로 걸러지지만, 예외는 컴파일은 잘 되지만 실행 중에 발생하게 된다.

> 예외처리는 어떻게 할까?

### try-catch문

try 블록 내에서 예외가 발생한 경우,

1. 예외가 발생한 이후 문장은 실행되지 않는다
2. 예외와 일치하는 catch 블록이 있다면 문장을 수행한다
3. finally 블록은 항상 실행된다(생략 가능)

```java
public void arrayOutOfBound() {
    int[] intArray = new int[5];
    
    try {
        System.out.println(intArray[5]);
        System.out.println("This code should run."); // 실행되지 않음, try블럭에서 예외가 발생할 경우, 예외 발생 위치 이후의 문장들은 수행되지 않는다.
    } catch (Exception e) {
        System.err.println(intArray.length);
    } finally {
        //예외여부와 관계없이 실행
        System.out.println("finally is called");
    }
    System.out.println("This code must run.");
}
```

```java
 > 5
 > finally is called
 > This code must run.
```

### 다중 catch블록

catch 블록은 순서를 따진다. Exception 클래스는 모든 예외 클래스의 상위 클래스이기 때문에 가장 아래에 놓여야 한다. 상위 클래스가 먼저 catch하면 모든 예외를 처리하게 되므로 다음 catch 블록 부터는 비교 수행 할 수 없기 때문이다.

## 예외의 종류
 
- checked Exception
  - 반드시 예외 처리를 해야한다. try-catch로 예외를 처리하거나 throw로 호출한 메소드에게 예외를 던져야 한다.
  - ex)IOException, SQLException, ClassNotFoundException
- unchecked Exception
  - 예외 발생을 미리 감지하지 못했을 때 발생한다. 컴파일러가 체크하지 않는 예외로 예외 처리를 하지 않아도 컴파일이 가능하다. 명시적인 예외 처리를 하지 않아도 된다.
  - ex)NullPointerException, IndexOutOfBoundException
- error
  - Java 프로그램 밖에서 발생한 예외이다. 디스크나 메인보드 고장이 여기에 속한다.

## java.lang.Throwable 클래스

Error, Exception의 상위 클래스로 모든 예외 클래스들이 가지고 있는 공통된 메소드를 정의

- getMessage() : 예외 메세지를 문자열로 반환
- toString() : 예외 메세지와 예외 클래스 이름을 제공
- printStackTrace() : 예외가 발생하게 된 메소드들의 호출 관계 출력

## 예외의 책임 전가 throws

throws로 선언되어 있는 메소드를 **호출한 메소드**에서 예외를 처리한다. 메소드 선언부 끝에 작성되어 예외를 호출한 곳으로 넘기는 역할을 한다.

